prefect_kubernetes/utilities.py

```diff
@@ -180,3 +185,126 @@ def _slugify_label_value(value: str, max_length: int = 63) -> str:
+
+        reconnect_exceptions = (
+            reconnect_exceptions
+            if reconnect_exceptions is not None
+            else self.DEFAULT_RECONNECT_EXCEPTIONS
+        )
+        self.reconnect_exceptions = tuple(reconnect_exceptions)
+
+    def _stream(self, cache: Optional[FIFOCache], func: Callable, *args, **kwargs):
+        """
+        A private method for streaming API objects or logs from a Kubernetes
+        client function. This method will reconnect the stream on certain
+        configurable exceptions and deduplicate results on reconnects if
+        streaming API objects.
+        """
+        keep_streaming = True
+        while keep_streaming:
+            try:
+                for event in self.watch.stream(func, *args, **kwargs):
+                    # check that we want to and can track this object
+                    if (
+                        cache is not None
+                        and isinstance(event, dict)
+                        and "object" in event
+                    ):
+                        uid = event["object"].metadata.uid
+                        if uid not in cache:
+                            cache[uid] = None
+                            yield event
+                    else:
+                        yield event
+                else:
+                    # Case: we've finished iterating
+                    keep_streaming = False
+            except self.reconnect_exceptions:
+                # Case: We've hit an exception we're willing to retry on
+                if self.logger:
+                    self.logger.error("Unable to connect, retrying...", exc_info=True)
+                time.sleep(1)
+            except Exception:
+                # Case: We hit an exception we're unwilling to retry on
+                if self.logger:
+                    self.logger.exception(
+                        f"Unexpected error while streaming {func.__name__}"
+                    )
+                keep_streaming = False
+                self.stop()
+                raise
+
+        self.stop()
+
+    def api_object_stream(self, func: Callable, *args, **kwargs):
+        """
+        Create a FIFOCache to maintain a record of API objects that have been
+        seen. This is useful because `_stream` will reconnect a stream on
+        `self.reconnect_exceptions` and on reconnect it will restart streaming all
+        objects. This cache prevents the same object from being yielded twice.
+
+        Args:
+            func: A Kubernetes client function to call which produces a stream of API o
+                bjects
+            *args: Positional arguments to pass to `func`
+            **kwargs: Keyword arguments to pass to `func`
+
+        Returns:
+            An iterator of API objects
+        """
+        cache = FIFOCache(maxsize=self.max_cache_size)
+        yield from self._stream(cache, func, *args, **kwargs)
+
+    def log_stream(self, func: Callable, *args, **kwargs):
```

Comment:
This function looks like it's a very thin wrapper around `_stream` and doesn't inherently have any knowledge of logs. Would we be able to remove this method and make `_stream` 'public' instead?

Reply:
I think that it'll be a good experience to have `stream` publicly available for a more general use case. I removed `log_stream` and updated the `stream` docstring to note it's behaviors on log stream client methods.